ethstorage/storage_manager.go

```diff
@@ -293,6 +302,143 @@ func (s *StorageManager) syncCheck(kvIdx uint64) error {
 	return nil
 }
 
+// DownloadAllMetas This function download the blob hashes of all the local storage shards from the smart contract
+func (s *StorageManager) DownloadAllMetas() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	header, err := s.l1Source.HeaderByNumber(context.Background(), big.NewInt(rpc.FinalizedBlockNumber.Int64()))
+	if err != nil {
+		return err
+	}
+	l1 := header.Number.Int64()
+
+	lastKvIdx, err := s.l1Source.GetStorageLastBlobIdx(l1)
+	if err != nil {
+		return err
+	}
+
+	for _, sid := range s.Shards() {
+		first, limit := s.KvEntries()*sid, s.KvEntries()*(sid+1)
+
+		// batch request metas until the lastKvIdx
+		end := limit
+		if end > lastKvIdx {
+			end = lastKvIdx
+		}
+		log.Info("Begin to download metas", "shard", sid, "first", first, "end", end, "limit", limit, "lastKvIdx", lastKvIdx)
+		ts := time.Now()
+
+		err := s.downloadMetaInParallel(first, end, l1)
+		if err != nil {
+			return err
+		}
+
+		log.Info("All the metas has been downloaded", "first", first, "end", end, "time", time.Since(ts).Seconds())
+		ts = time.Now()
+
+		// empty blobs
+		for i := end; i < limit; i++ {
+			meta := [32]byte{}
+			new(big.Int).SetInt64(int64(i)).FillBytes(meta[0:5])
+
+			s.blobMetas[i] = meta
+		}
+
+		log.Info("Empty metas has been filled", "first", end, "limit", limit, "time", time.Since(ts).Seconds())
+	}
+
+	return nil
+}
+
+func (s *StorageManager) downloadMetaInParallel(from, to uint64, l1 int64) error {
+	var wg sync.WaitGroup
+	taskNum := uint64(MetaDownloadThread)
+
+	// We don't need to download in parallel if the meta amount is small
+	if to - from < uint64(taskNum) * MetaBatchSize {
+		return s.downloadMetaInRange(from, to, l1, 0)
+	}
+
+	chanRes := make(chan error, taskNum)
+	defer close(chanRes)
+
+	rangeSize := (to - from) / uint64(taskNum)
+	for taskIdx := uint64(0); taskIdx < taskNum; taskIdx++ {
+		rangeStart := taskIdx * rangeSize
+		rangeEnd := (taskIdx + 1) * rangeSize
+		if taskIdx == taskNum - 1 {
+			rangeEnd = to
+		}
+		wg.Add(1)
+
+		go func(start, end, taskId uint64, out chan<- error) {
+			defer wg.Done()
+			err := s.downloadMetaInRange(start, end, l1, taskId)
+
+			chanRes <- err
+		}(rangeStart, rangeEnd, taskIdx, chanRes)
+	}
+
+	wg.Wait()
+
+	for i := uint64(0); i < taskNum; i++ {
+		res := <- chanRes
+		if (res != nil) {
+			return res
+		}
+	}
+
+	return nil
+}
+
+func (s *StorageManager) downloadMetaInRange(from, to uint64, l1 int64, taskId uint64) error {
+	for from < to {
+		batchLimit := from + MetaBatchSize
+		if batchLimit > to {
+			batchLimit = to
+		}
+		kvIndices := []uint64{}
+		for i := from; i < batchLimit; i++ {
+			kvIndices = append(kvIndices, i)
+		}
+
+		metas, err := s.l1Source.GetKvMetas(kvIndices, l1)
+		if err != nil {
+			return err
+		}
+
+		s.metaMapMu.Lock()
```

Comment:
I guess you can reuse mu without introducing metaMapMu.  Further, it should check whether l1 has changed when writing it into the map to avoid race conditions.